### DISCUSSION TITLE
Client controlled nullability: inconsistency of assertions?

### DISCUSSION THREAD

**Discussion by benjie** [graphql/graphql-wg#1009] [General]
@twof Now that CCN doesn't simply change the nullability of a field, the intent of how certain cases should be handled is less clear to me. The [current spec edits](https://github.com/graphql/graphql-spec/pull/895/files) state:

> In addition to being `Non-Nullable`, if a field marked with `!` resolves to `null`, it propagates to the nearest parent field marked with a `?` or to `data` if there is not a parent marked with a `?`. An error is added to the `errors` array identical to if the field had been `Non-Nullable` in the schema.

Consider the following:

```graphql
type Query { a: A }
type A { b: B }
type B { c: Int }

query {
  a? {
    b {
      c!
    }
  }
}
```

Should `a.b.c` be null, then the result should be `{a: null}` due to the throw/catch nature of `!` / `?`.

However, if `a.b` is null the spec seems to indicate that `{a:{b: null}}` acceptable even though `a.b.c` is not present. It seems confusing that if this is acceptable, why is it not also an acceptable solution to the situation where `a.b.c` alone was null?

Previously we've discussed the idea that `!` might make all field from the nearest `?` (or the root) to the `!` field non-nullable also - i.e. `!` would guarantee that if the container (`?` field, or root) is present, then the full path exists, gaining consistency. Maybe this would be a good solution to this seeming inconsistency?

Personally, I'd be happy if `!` and `?` went back to just setting the effective nullability of the field and having no further behaviours associated.

**Comment by twof** [comment]
@benjie I think you're right. If a query author doesn't believe their feature works when `c` is missing, then they should also reject responses with missing `b` since that means `c` would also be missing.

@captbaritone @josephsavona does it still work with Relay's fragment isolation plan if all fields between `!` and `?` are non-null?

**Reply by twof** [reply]
@benjie 
Talking to @mjmahone and co (feel free to tag them as well. Idk their handles) at GraphQL Conf, this query was constructed as a counterexample where having every field between a non-null assert and catch be non-nullable would be harmful.

```graphql
{
  me? {
    bestfriend {
      name!
    }
  }
}
```
In this case, you only want the propagation to happen if `name` is null, but not if `bestfriend` is null. If `bestfriend` is missing, then we don't care that `name` is also missing, and we're fine if `bestfriend` is missing.

**Reply by josephsavona** [reply]
Yeah, this is a great example of why `!` doesn't imply that parents are also non-nullable. In @twof's example, the intent is "if and only if `bestfriend` is present, then `name` must be present too". To indicate that `bestfriend` must also be present, you'd mark as such with `!`. Either of the following is reasonable depending on your requirements:

```
// bestfriend is optional, but if present it must have a name
me? {
  bestfriend {
    name!
  }
}

// bestfriend must be present and must have a name
me? {
  bestfriend! {
    name!
  }
}
```
cc @mjmahone, @captbaritone, @mofeiZ, @davidmccabe

**Reply by josephsavona** [reply]
This approach also makes a lot of sense if you think about named fragment spreads (separate but complimentary to this proposal). Named fragment spreads provide (among other things) a potential boundary point for required data. It would be very useful to write something like:

```
query UserProfile($id: ID!) {
  user(id: $id) {
    name
    ...Header
    Timeline: ...Timeline # named fragment spreads
  }
}

fragment Timeline on User {
  posts! { # can't render timeline without posts
    # ... 
  }
}
```

If `!` implicitly made all parent fields non-nullable, this would defeat the point of named fragment spreads as boundary. For example if posts was missing, the whole query would be nulled out. Whereas the proposed behavior lets the `Timeline` alias act as an (implicitly) nullable boundary. If posts is null at runtime, then only `data.user.Timeline` gets nulled out. To clarify, the proposed behavior is fully motivated by normal use within fragments. But it *also* aligns with named fragment spreads.

**Reply by benjie** [reply]
So what you're saying with this query:

```graphql
me? {
  bestfriend {
    name!
  }
  avatar
}
```

is:

- Tell me my avatar and my bestfriend's name
- If I don't have a bestfriend, just tell me my avatar, that's fine
- If I have a bestfriend but they don't have a name, then **don't even tell me my avatar - just null me out entirely**

It's this last bullet that feels wrong to me - it's a divergence from the current non-null behaviour of GraphQL and I can't quite wrap my head around what makes it beneficial versus instead returning `{data:{me:{bestfriend: null, avatar: "/avatar.png"}},errors: [...]}`

> In @twof's example, the intent is "if and only if bestfriend is present, then name must be present too". To indicate that bestfriend must also be present, you'd mark as such with !

This is exactly what would happen if `!` made the field non-null, rather than being a "throw"-like behaviour as currently suggested.

> If ! implicitly made all parent fields non-nullable, this would defeat the point of named fragment spreads as boundary. 

You could (and should?) make the named fragment spread behave like it has a `?` so I'm not sure this is really a problem.

---

Do you have a more convincing use case you can share where it makes it clear _why_ `me` should be nulled out rather than the (already nullable) `me.bestfriend` field being nulled out if `name` is null?

- Is it to "swallow" errors, so that the rest of the query can still be stored to cache even if that part's not okay?
- Is it something to do with fragment modularity? If so, how does just effectively changing the type to GraphQLNonNull(...) not solve it in the same way?
- Is it to give a stricter indication of consistency? E.g. we _know_ you have a best friend, but the data is somehow invalid so they'd be nulled out, so rendering it in this broken state would be super confusing so let's just null out this whole subtree?
- Is it some other reason I'm missing?

**Reply by mofeiZ** [reply]
Chatted with @captbaritone and I believe our internal use case for client-controlled nullability / @required fits with # 3.

> Give a stricter indication of consistency? E.g. we know you have a best friend, but the data is somehow invalid so they'd be nulled out, so rendering it in this broken state would be super confusing so let's just null out this whole subtree

As an example (hopefully not too verbose!), we might have the query
```
me? {
  bestfriend {
    name!
    avatar
  }
  avatar
}
```
In the UI / Component tree, we have
- UserProfileComponent
  - AvatarComponent
  - BestFriendProfileComponent
    - AvatarComponent

`BestFriendProfileComponent` might not know how to render without a name. Since this is an error (according to the business logic of BestFriendProfileComponent), bestfriend should *not* be nulled out. Instead, this error should bubble up to the nearest error-handling boundary (me), and we could display an error fallback in place of a User Profile. 
However, it's perfectly valid for a user to not have a best friend.

I'm still fairly new to this proposal, so please feel free to correct if any of ^ is incorrect!

**Reply by benjie** [reply]
Thanks for discussing with @captbaritone and sharing the result, @mofeiZ. My concern is that this can lead to less-than-expected outcomes. Take your example, let's imagine that `avatar` is non-nullable and yet it is presented with a null. It would throw, and the nullable `bestfriend` field would become null (adding an error to the errors list) and thus the `name` field would no longer exist.

Is this acceptable in your situation because this is a concrete error, rather than a self-imposed issue from `!`? (I've heard Relay doesn't do much with responses that have `errors`?)

If you added `!` to avatar, does it make sense that the error would be handled differently now (i.e. caught by the `?`)? Actually... _would_ it be handled differently now, or would the underlying non-null win out over the assertion `!`?

**Reply by twof** [reply]
@josephsavona @benjie @mofeiZ @mjmahone  Hello! I'm hopping back on these discussion threads to try and see what I can do to move things forward. After reading back and forth here, I think I largely agree with the arguments that @benjie has made, and I'm leaning towards making everything between an assertion and a catch operator a required field. 

With the amount of time that's passed, I've lost a bit of context and I'm having trouble remembering if the folks I was talking to from Meta felt that was a no-go for Relay, or if they just felt that it wasn't preferable. Do the folks on the other side of the debate still feel strongly about this? If so would you like to hop on a call so we can hash this out?

**Reply by LunaticMuch** [reply]
> I've lost a bit of context 

It's a key point. I am late at this party 🙇🏻 so reading it with a fresh perspective I end with:

1. it feels a bit overengineering a problem which is not 100% crystal clear (or it's party included in [this section from Relay](https://relay.dev/docs/guides/required-directive/#why-not-implement-this-at-the-schemaserver-level))
2. while a CCN can have some interesting benefits, why shall it be also include additional behaviours (clearly odd in some cases?)

A lot in GraphQL is with the contract and setting the contract right. Adding CCN with additional behaviours seems to invalidate this basic principle.

**Reply by AnthonyMDev** [reply]
> BestFriendProfileComponent might not know how to render without a name. Since this is an error (according to the business logic of BestFriendProfileComponent), bestfriend should not be nulled out. Instead, this error should bubble up to the nearest error-handling boundary (me), and we could display an error fallback in place of a User Profile.
However, it's perfectly valid for a user to not have a best friend.

I agree with @benjie. This seems to lead to unintuitive situations in a lot of cases. I'm strongly in favor of errors only bubbling up to the next nullable parent (whether that parent is nullable in the schema or by use of `?`).

@mofeiZ It feels like what you want here is too closely tied to client specific behaviors, but it feels like it would make CCN become a lot more than CCN (as @LunaticMuch) mentioned. Making the `?` operator an explicit error boundary makes CCN a lot more complicated to implement and for client developers to use and reason about.

--- 

Please let me know if I'm misunderstanding something about what you're goals are here in this example.

In your example, I'm not convinced that making the `user` object `null` is actually the best case scenario either. Ideally, you would likely still want to get the user's `avatar` and only display an error where the `bestfriend` component would be (instead of for the entire user profile).

I think the struggle here is that the proposed response data does not give you a good understanding of _why_ the `bestfriend` field is `null`.

`bestfriend` could be `null` because:
1. The user has no best friend
2. The required `name!` field was null, so the best friend object was invalid.

While there is no indication of which of these is the cause at the site of the `bestfriend` field, that information can still be obtained by looking for the CCN error for the `name` field. 

This isn't always easy to do though.

Firstly, view components usually don't have a reference to the entire GraphQL request and it's errors, so you'd need to build something to pass that around, this could get messy and detract from code quality. Additionally, if there are many fields that have `!` CCN annotations, you would need to look through all of the errors and see if _any_ of them indicate that the `bestfriend` field should be `null` due to an `null` child field.

---

Thinking through what the client would need to do to glean this information does make me think there is something missing here still. However I'm not convinced that making the `?` the explicit error boundary is the right choice. I am more inclined to look for how we can better indicate the necessary information to the client when a CCN error occurs.

Overall, the issue seems to be more about being able to understand _why_ the `bestfriend` field is `null`.

Even though this may be a bit messy, I would argue that this is a concern of the client. All of the information necessary to deduce if this `null` object is really `null` or due to a CCN error is provided in the response errors. Constructing some client-side infrastructure that parses the errors and then propagates information about them to the consuming models/components is definitely possible.

**Reply by AnthonyMDev** [reply]
> After reading back and forth here, I think I largely agree with the arguments that @benjie has made, and I'm leaning towards making everything between an assertion and a catch operator a required field.

@twof Maybe I'm missing something, but it sounds like that is the opposite of what @benjie is in favor of?

**Reply by twof** [reply]
@AnthonyMDev Yeah you're right. You can read more of Benjie's thoughts here. https://github.com/graphql/client-controlled-nullability-wg/issues/2#issuecomment-1647383722

We're going to be discussing this at the CCN WG meeting on the 26th, and again at the main WG meeting on August 3rd if you'd like to be a part of those discussions. https://github.com/graphql/client-controlled-nullability-wg/blob/main/agendas/2023/2023-07-26.md

**Reply by AnthonyMDev** [reply]
I'll be at both meetings :)

**Comment by IvanGoncharov** [comment]
I agree with @benjie that the semantics of error boundaries is strange.
As I understand it originated from the relay's `@required` directives and was boosted by the observation that in the absolute majority of cases clients just used the `THROW` argument.
https://relay.dev/docs/guides/required-directive/#throw-unrecoverable

My point is that Relay's `@required(action: THROW)` makes a lot of sense since it uses existing __react error boundaries__ that will handle any exception and not only GraphQL ones.
That behavior exactly matches the business use case suggested here https://github.com/graphql/graphql-wg/discussions/1009#discussioncomment-2940588

However, I don't think we should create parallel error boundaries (parallel to client code, e.g. react error boundaries) in GraphQL itself I think correct relay behavior is optimal and we should preserve it.

**Reply by AnthonyMDev** [reply]
Yes, there is no reason why adoption of CCN into the spec prevents Relay from continuing to use the `@required` directive, right?

My understanding is that in the original proposal, the `?` operator really only became an error boundary as a welcome side-effect of the error boundary being the next nullable field, and that could make a field nullable.

If the `?` _must_ be used for an error boundary to exist at all, then this really is much more than just "CCN". You would need to use the `?` operator even if the field you wanted to be the error boundary is already nullable.

If we are going to explicitly include syntax for setting error boundaries, I'd suggest that we explore options for different syntax. Because it feels like we are overloading the functionality of the `?` operator. I would say we could start proposing syntax for defining an explicit error boundary as an additional RFC. It doesn't seem necessary that we include that functionality for CCN to be implemented.

**Comment by benjie** [comment]
I wrote my concerns up in more detail here: https://github.com/graphql/client-controlled-nullability-wg/issues/2#issuecomment-1647383722
